<html lang="en">

  <head>
    <META content="text/html; charset=UTF-8" http-equiv="Content-Type"/>
    <meta charset="UTF-8"/>
    <meta content="width=device-width, initial-scale=1" name="viewport"/>
    <title>Jvarkit : Java utilities for Bioinformatics</title>
    <link href="/jvarkit/assets/css/style.css" rel="stylesheet"/>
  </head>

<body>
<div class="container markdown-body">
<h1>Jvarkit : Java utilities for Bioinformatics</h1>
<a href="https://travis-ci.org/lindenb/jvarkit"><img src="https://travis-ci.org/lindenb/jvarkit.svg"/></a><br/>

<h2>Tools</h2>
<table>

<thead>
	
<tr>
		
		<th>Name</th>
		<th>Description</th>
		<th>Keywords</th>
		<th>Terms</th>
		<th>Misc</th>
	
</tr>

</thead>

<tbody>

<tr id="samjs"><th><a href="SamJavascript.html" title="SamJavascript">samjs</a></th><td>Filters a BAM using a javascript expression ( java nashorn engine  ).</td><td>sam bam nashorn javascript filter</td><td/><td/></tr>
<tr id="casectrljfx"><th><a href="CaseControlJfx.html" title="CaseControlJfx">casectrljfx</a></th><td>display jfx chart of case/control maf from a VCF and a pedigree</td><td>vcf pedigree case control visualization jfx chart maf</td><td/><td/></tr>
<tr id="pubmedcodinglang"><th><a href="PubmedCodingLanguages.html" title="PubmedCodingLanguages">pubmedcodinglang</a></th><td>Programming language use distribution from recent programs / articles</td><td>pubmed xml code programming</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcffixindels"><th><a href="VCFFixIndels.html" title="VCFFixIndels">vcffixindels</a></th><td>Fix samtools indels (for @SolenaLS)</td><td>vcf indel</td><td/><td/></tr>
<tr id="vcftail"><th><a href="VcfTail.html" title="VcfTail">vcftail</a></th><td>print the last variants of a vcf</td><td>vcf</td><td/><td/></tr>
<tr id="vcfhead"><th><a href="VcfHead.html" title="VcfHead">vcfhead</a></th><td>print the first variants of a vcf</td><td>vcf</td><td/><td/></tr>
<tr id="vcfmulti2oneallele"><th><a href="VcfMultiToOneAllele.html" title="VcfMultiToOneAllele">vcfmulti2oneallele</a></th><td>'one variant with N ALT alleles' to 'N variants with one ALT'</td><td>vcf</td><td/><td/></tr>
<tr id="vcfin"><th><a href="VcfIn.html" title="VcfIn">vcfin</a></th><td>Only prints variants that are contained/not contained into another VCF</td><td>vcf compare</td><td/><td/></tr>
<tr id="vcfpredictions"><th><a href="VCFPredictions.html" title="VCFPredictions">vcfpredictions</a></th><td>Basic Variant Effect prediction using ucsc-known gene</td><td>vcf annotation prediction</td><td/><td/></tr>
<tr id="vcffilterso"><th><a href="VcfFilterSequenceOntology.html" title="VcfFilterSequenceOntology">vcffilterso</a></th><td>Filter a VCF file annotated with SNPEff or VEP with terms from Sequence-Ontology. Reasoning : Children of user's SO-terms will be also used.&lt;</td><td>vcf filter sequenceontology prediction so</td><td/><td/></tr>
<tr id="vcffilterjs"><th><a href="VCFFilterJS.html" title="VCFFilterJS">vcffilterjs</a></th><td>Filtering VCF with javascript expressions</td><td>vcf filter javascript json nashorn</td><td>[ID:0000003]i_use_it_regularly</td><td/></tr>
<tr id="vcfpolyx"><th><a href="VCFPolyX.html" title="VCFPolyX">vcfpolyx</a></th><td>Number of repeated REF bases around POS.</td><td>vcf repeat</td><td/><td/></tr>
<tr id="vcfbed"><th><a href="VCFBed.html" title="VCFBed">vcfbed</a></th><td>Transfer information from a BED to a VCF</td><td>bed vcf annotation</td><td/><td/></tr>
<tr id="vcftrio"><th><a href="VCFTrios.html" title="VCFTrios">vcftrio</a></th><td>Find mendelian incompatibilitie in a VCF</td><td>vcf mendelian pedigree</td><td/><td/></tr>
<tr id="groupbygene"><th><a href="GroupByGene.html" title="GroupByGene">groupbygene</a></th><td>Group VCF data by gene/transcript. By default it uses data from VEP , SnpEff</td><td>vcf gene</td><td/><td/></tr>
<tr id="addlinearindextobed"><th><a href="AddLinearIndexToBed.html" title="AddLinearIndexToBed">addlinearindextobed</a></th><td>Use a Sequence dictionary to create a linear index for a BED file. Can be used as a X-Axis for a chart.</td><td>bed reference</td><td/><td/></tr>
<tr id="almostsortedvcf"><th><a href="AlmostSortedVcf.html" title="AlmostSortedVcf">almostsortedvcf</a></th><td>Sort an 'almost' sorted VCF. Most variants should be sorted but a few  consecutive lines might have been switched by a caller.</td><td/><td/><td/></tr>
<tr id="backlocate"><th><a href="BackLocate.html" title="BackLocate">backlocate</a></th><td>Mapping a mutation on a protein back to the genome.</td><td>vcf annotation prediction protein</td><td/><td/></tr>
<tr id="bam2fastq"><th><a href="BamToFastq.html" title="BamToFastq">bam2fastq</a></th><td>Same as picard/SamToFastq but allow missing reads + shuffle reads using hash(name) so you can use them with bwa. </td><td>fastq</td><td>[ID:0000005]unmaintained</td><td><span style="color:orange;" title="deprecated">use picard</span>. </td></tr>
<tr id="bam2raster"><th><a href="Bam2Raster.html" title="Bam2Raster">bam2raster</a></th><td>BAM to raster graphics</td><td>bam alignment graphics visualization png</td><td/><td/></tr>
<tr id="bam2svg"><th><a href="BamToSVG.html" title="BamToSVG">bam2svg</a></th><td>BAM to raster graphics</td><td>bam alignment graphics visualization svg</td><td/><td/></tr>
<tr id="bam2xml"><th><a href="Bam2Xml.html" title="Bam2Xml">bam2xml</a></th><td>converts a BAM to XML</td><td>sam bam xml</td><td/><td/></tr>
<tr id="bam2wig"><th><a href="Bam2Wig.html" title="Bam2Wig">bam2wig</a></th><td>Bam to fixedStep Wiggle converter. Parses the cigar String to get the depth. Memory intensive: must alloc sizeof(int)*size(chrom)</td><td>bam wig wiggle</td><td/><td/></tr>
<tr id="bamcmpcoverage"><th><a href="BamCmpCoverage.html" title="BamCmpCoverage">bamcmpcoverage</a></th><td>Creates the figure of a comparative view of the depths sample vs sample. Memory consideration: the tool alloc an array of bits which size is: (MIN(maxdepth-mindepth,pixel_width_for_one_sample) * count_samples)^2</td><td>sam bam visualization coverage</td><td/><td/></tr>
<tr id="vcfgo"><th><a href="VcfGeneOntology.html" title="VcfGeneOntology">vcfgo</a></th><td>Find the GO terms for VCF annotated with SNPEFF or VEP</td><td>vcf go</td><td/><td/></tr>
<tr id="bamrenamechr"><th><a href="ConvertBamChromosomes.html" title="ConvertBamChromosomes">bamrenamechr</a></th><td>Convert the names of the chromosomes in a BAM file</td><td>sam bam chromosome contig</td><td/><td/></tr>
<tr id="bamstats04"><th><a href="BamStats04.html" title="BamStats04">bamstats04</a></th><td>Coverage statistics for a BED file.</td><td>bam coverage statistics bed</td><td/><td/></tr>
<tr id="bamstats05"><th><a href="BamStats05.html" title="BamStats05">bamstats05</a></th><td>Coverage statistics for a BED file, group by gene</td><td>bam coverage statistics bed</td><td/><td/></tr>
<tr id="batchigvpictures"><th><a href="BatchIGVPictures.html" title="BatchIGVPictures">batchigvpictures</a></th><td/><td/><td/><td/></tr>
<tr id="bedliftover"><th><a href="BedLiftOver.html" title="BedLiftOver">bedliftover</a></th><td>Lift-over a VCF file</td><td>bed liftover</td><td/><td/></tr>
<tr id="bedrenamechr"><th><a href="ConvertBedChromosomes.html" title="ConvertBedChromosomes">bedrenamechr</a></th><td>Convert the names of the chromosomes in a Bed file</td><td>bed chromosome contig convert</td><td/><td/></tr>
<tr id="biostar103303"><th><a href="Biostar103303.html" title="Biostar103303">biostar103303</a></th><td>Calculate Percent Spliced In (PSI).</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar130456"><th><a href="Biostar130456.html" title="Biostar130456">biostar130456</a></th><td>Individual VCF files from main VCF file</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar59647"><th><a href="Biostar59647.html" title="Biostar59647">biostar59647</a></th><td>SAM/BAM to XML</td><td/><td/><td/></tr>
<tr id="biostar76892"><th><a href="Biostar76892.html" title="Biostar76892">biostar76892</a></th><td>fix strand of two paired reads close but on the same strand. </td><td>sam bam</td><td/><td/></tr>
<tr id="biostar77288"><th><a href="Biostar77288.html" title="Biostar77288">biostar77288</a></th><td>Low resolution sequence alignment visualization</td><td>bam sam visualization svg alignment</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar77828"><th><a href="Biostar77828.html" title="Biostar77828">biostar77828</a></th><td>Divide the human genome among X cores, taking into account gaps</td><td>workflow reference parallel</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar78285"><th><a href="Biostar78285.html" title="Biostar78285">biostar78285</a></th><td>Extract regions of genome that have 0 coverage See http://www.biostars.org/p/78285/</td><td>sam bam depth coverage</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar78400"><th><a href="Biostar78400.html" title="Biostar78400">biostar78400</a></th><td>add the read group info to the sam file on a per lane basis</td><td>sam bam xml read-group</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar81455"><th><a href="Biostar81455.html" title="Biostar81455">biostar81455</a></th><td>Defining precisely the genomic context based on a position .</td><td>bed gene knownGene ucsc</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar84452"><th><a href="Biostar84452.html" title="Biostar84452">biostar84452</a></th><td>remove clipped bases from BAM</td><td>sam bam clip</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar84786"><th><a href="Biostar84786.html" title="Biostar84786">biostar84786</a></th><td>Matrix transposition</td><td>matrix util</td><td/><td/></tr>
<tr id="biostar86363"><th><a href="Biostar86363.html" title="Biostar86363">biostar86363</a></th><td>Set genotype of specific sample/genotype comb to unknown in multisample vcf file. See http://www.biostars.org/p/86363/</td><td>sample genotype vcf</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar86480"><th><a href="Biostar86480.html" title="Biostar86480">biostar86480</a></th><td>Genomic restriction finder</td><td>rebase genome enzyme restricion genome</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar90204"><th><a href="Biostar90204.html" title="Biostar90204">biostar90204</a></th><td>Bam version of linux split. See also http://www.biostars.org/p/90204/</td><td>sam bam split util</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="msa2vcf"><th><a href="MsaToVcf.html" title="MsaToVcf">msa2vcf</a></th><td>Getting a VCF file from a CLUSTAW or a FASTA alignment. </td><td/><td/><td><span style="color:orange;" title="deprecated">use https://github.com/sanger-pathogens/snp_sites</span>. </td></tr>
<tr id="biostar95652"><th><a href="Biostar95652.html" title="Biostar95652">biostar95652</a></th><td>Drawing a schematic genomic context tree.</td><td>genbank svg tree evolution</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar139647"><th><a href="Biostar139647.html" title="Biostar139647">biostar139647</a></th><td>Convert alignment in Fasta/Clustal format to SAM/BAM file</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar145820"><th><a href="Biostar145820.html" title="Biostar145820">biostar145820</a></th><td>subsample BAM to fixed number of alignments.</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="blastmapannots"><th><a href="BlastMapAnnotations.html" title="BlastMapAnnotations">blastmapannots</a></th><td>Maps uniprot/genbank annotations on a blast result.</td><td>blast annotation genbank uniprot</td><td/><td/></tr>
<tr id="buildwpontology"><th><a href="BuildWikipediaOntology.html" title="BuildWikipediaOntology">buildwpontology</a></th><td>Build a simple RDFS/XML ontology from the Wikipedia Categories</td><td>wikipedia ontology rdf gui</td><td/><td/></tr>
<tr id="bwamemdigest"><th><a href="BWAMemDigest.html" title="BWAMemDigest">bwamemdigest</a></th><td/><td/><td/><td/></tr>
<tr id="bwamemnop"><th><a href="BWAMemNOp.html" title="BWAMemNOp">bwamemnop</a></th><td>Merge the other BWA-MEM alignements with its SA:Z:* attributes to an alignment containing a cigar string with 'N' (  Skipped region from the reference.)</td><td>bwa sam bam</td><td/><td/></tr>
<tr id="cmpbams"><th><a href="CompareBams.html" title="CompareBams">cmpbams</a></th><td>Compare two or more BAM files</td><td>sam bam compare</td><td/><td/></tr>
<tr id="cmpbamsandbuild"><th><a href="CompareBamAndBuild.html" title="CompareBamAndBuild">cmpbamsandbuild</a></th><td>Compare two  BAM files mapped on two different builds. Requires a liftover chain file</td><td/><td/><td/></tr>
<tr id="coveragenormalizer"><th><a href="CoverageNormalizer.html" title="CoverageNormalizer">coveragenormalizer</a></th><td>normalize BAM coverage</td><td/><td/><td/></tr>
<tr id="downsamplevcf"><th><a href="DownSampleVcf.html" title="DownSampleVcf">downsamplevcf</a></th><td>DownSample a VCF</td><td>vcf</td><td/><td/></tr>
<tr id="evs2bed"><th><a href="DumpExomeVariantServerData.html" title="DumpExomeVariantServerData">evs2bed</a></th><td>Download data from EVS http://evs.gs.washington.edu/EVS as a BED chrom/start/end/XML For later use, see VCFTabixml.</td><td/><td/><td/></tr>
<tr id="evs2vcf"><th><a href="EvsToVcf.html" title="EvsToVcf">evs2vcf</a></th><td>Download data from EVS http://evs.gs.washington.edu/EVS as a BED chrom/start/end/XML For later use, see VCFTabixml.</td><td/><td/><td/></tr>
<tr id="evsdumpxml"><th><a href="EvsDumpXml.html" title="EvsDumpXml">evsdumpxml</a></th><td>Download data from EVS http://evs.gs.washington.edu/EVS as XML file.</td><td/><td/><td/></tr>
<tr id="fastq2fasta"><th><a href="FastqToFasta.html" title="FastqToFasta">fastq2fasta</a></th><td>fastq -&gt; fasta</td><td>fastq fasta</td><td/><td><span style="color:orange;" title="deprecated">use awk, samtools...</span>. </td></tr>
<tr id="kg2bed"><th><a href="KnownGenesToBed.html" title="KnownGenesToBed">kg2bed</a></th><td>converts UCSC knownGenes file to BED.</td><td>ucsc bed knownGenes</td><td/><td/></tr>
<tr id="fastqentropy"><th><a href="FastqEntropy.html" title="FastqEntropy">fastqentropy</a></th><td>Compute the Entropy of a Fastq file (distribution of the length(gzipped(sequence))</td><td>fastq</td><td>[ID:0000005]unmaintained</td><td/></tr>
<tr id="fastqgrep"><th><a href="FastqGrep.html" title="FastqGrep">fastqgrep</a></th><td>Grep reads names in fastq</td><td>fastq</td><td/><td><span style="color:orange;" title="deprecated">use picard</span>. </td></tr>
<tr id="fastqjs"><th><a href="FastqJavascript.html" title="FastqJavascript">fastqjs</a></th><td>Filters a FASTQ file using javascript( java nashorn engine). </td><td>fastq</td><td>[ID:0000005]unmaintained</td><td/></tr>
<tr id="fastqphred64to33"><th><a href="ConvertPhred64toFastq33.html" title="ConvertPhred64toFastq33">fastqphred64to33</a></th><td>Convert Illumina Fastq 64 encoding to Fastq 33</td><td>fastq</td><td>[ID:0000005]unmaintained</td><td/></tr>
<tr id="fastqrevcomp"><th><a href="FastqRevComp.html" title="FastqRevComp">fastqrevcomp</a></th><td>produces a reverse-complement fastq (for mate pair alignment see http://seqanswers.com/forums/showthread.php?t=5085 )</td><td>fastq</td><td/><td/></tr>
<tr id="fastqshuffle"><th><a href="FastqShuffle.html" title="FastqShuffle">fastqshuffle</a></th><td>Shuffle Fastq files</td><td>fastq</td><td>[ID:0000005]unmaintained</td><td/></tr>
<tr id="fastqsplitinterleaved"><th><a href="FastqSplitInterleaved.html" title="FastqSplitInterleaved">fastqsplitinterleaved</a></th><td>Split interleaved Fastq files.</td><td>fastq</td><td>[ID:0000005]unmaintained</td><td/></tr>
<tr id="findallcoverageatposition"><th><a href="FindAllCoverageAtPosition.html" title="FindAllCoverageAtPosition">findallcoverageatposition</a></th><td>Find depth at specific position in a list of BAM files. My colleague Estelle asked: in all the BAM we sequenced, can you give me the depth at a given position ?</td><td>bam coverage search depth</td><td/><td/></tr>
<tr id="findavariation"><th><a href="FindAVariation.html" title="FindAVariation">findavariation</a></th><td>Finds a specific mutation in a list of VCF files</td><td>vcf variation search</td><td/><td/></tr>
<tr id="findcorruptedfiles"><th><a href="FindCorruptedFiles.html" title="FindCorruptedFiles">findcorruptedfiles</a></th><td>Reads filename from stdin and prints corrupted NGS files (VCF/BAM/FASTQ/BED/TBI/BAI)</td><td>vcf bam fastq bed</td><td/><td/></tr>
<tr id="findmyvirus"><th><a href="FindMyVirus.html" title="FindMyVirus">findmyvirus</a></th><td>Find my Virus. Created for Adrien Inserm. Proportion of reads mapped on HOST/VIRUS. </td><td/><td/><td/></tr>
<tr id="findnewsplicesites"><th><a href="FindNewSpliceSites.html" title="FindNewSpliceSites">findnewsplicesites</a></th><td>use the 'N' operator in the cigar string to find unknown splice sites</td><td>rnaseq splice</td><td>[ID:0000008]it_was_a_one_shot</td><td/></tr>
<tr id="fixvarscanmissingheader "><th><a href="FixVarScanMissingVCFHeader.html" title="FixVarScanMissingVCFHeader">fixvarscanmissingheader </a></th><td>Fix the sample name in the #CHROM header and fix VCF header missingh in varscan2. Sometimes, but not always, it happends. See  http://seqanswers.com/forums/showthread.php?t=33235 </td><td/><td/><td/></tr>
<tr id="fixvcf"><th><a href="FixVCF.html" title="FixVCF">fixvcf</a></th><td>Fix a VCF if INFO or FILTER are missing</td><td/><td/><td/></tr>
<tr id="fixvcfformat"><th><a href="FixVcfFormat.html" title="FixVcfFormat">fixvcfformat</a></th><td>Fix PL format in VCF. Problem is described in http://gatkforums.broadinstitute.org/discussion/3453</td><td/><td/><td/></tr>
<tr id="fixvcfmissinggenotypes"><th><a href="FixVcfMissingGenotypes.html" title="FixVcfMissingGenotypes">fixvcfmissinggenotypes</a></th><td>After a VCF-merge, read a VCF, look back at some BAMS to tells if the missing genotypes were homozygotes-ref or not-called. If the number of reads is greater than min.depth, then the missing genotypes is said hom-ref.</td><td>sam bam vcf</td><td/><td/></tr>
<tr id="gcpercentanddepth"><th><a href="GcPercentAndDepth.html" title="GcPercentAndDepth">gcpercentanddepth</a></th><td>Extracts GC% and depth for multiple bam using a sliding window</td><td>gc% depth coverage</td><td/><td/></tr>
<tr id="genomicjaspar"><th><a href="GenomicJaspar.html" title="GenomicJaspar">genomicjaspar</a></th><td>Find jaspar patterns in FASTA sequences. Reports a BED file.</td><td>jaspar genomic pattern</td><td/><td/></tr>
<tr id="howmanybamdict"><th><a href="HowManyBamDict.html" title="HowManyBamDict">howmanybamdict</a></th><td>finds if there's are some differences in the sequence dictionaries.</td><td>sam bam dict</td><td/><td/></tr>
<tr id="illuminadir"><th><a href="IlluminaDirectory.html" title="IlluminaDirectory">illuminadir</a></th><td>Create a structured (**JSON** or **XML**) representation of a directory containing some Illumina FASTQs.</td><td>json xml illumina fastq</td><td/><td/></tr>
<tr id="ilmnfastqstats"><th><a href="IlluminaStatsFastq.html" title="IlluminaStatsFastq">ilmnfastqstats</a></th><td>Reads filenames from stdin: Count FASTQs in Illumina Result.</td><td/><td/><td/></tr>
<tr id="impactofduplicates"><th><a href="ImpactOfDuplicates.html" title="ImpactOfDuplicates">impactofduplicates</a></th><td>Impact of Duplicates per BAM.</td><td>bam</td><td/><td/></tr>
<tr id="liftover2svg"><th><a href="LiftOverToSVG.html" title="LiftOverToSVG">liftover2svg</a></th><td>Convert LiftOver chain files to animated SVG</td><td>svg liftover ucsc xml</td><td/><td/></tr>
<tr id="mapuniprot"><th><a href="MapUniProtFeatures.html" title="MapUniProtFeatures">mapuniprot</a></th><td>map uniprot features on reference genome</td><td>uniprot bed fasta reference xjc xml</td><td/><td/></tr>
<tr id="mergesplittedblast"><th><a href="MergeSplittedBlast.html" title="MergeSplittedBlast">mergesplittedblast</a></th><td>merge blast Hits from splitted BLAST database</td><td>blast</td><td/><td/></tr>
<tr id="ncbitaxonomy2xml"><th><a href="NcbiTaxonomyToXml.html" title="NcbiTaxonomyToXml">ncbitaxonomy2xml</a></th><td>Dump NCBI taxonomy tree as a hierarchical XML document</td><td>taxonomy ncbi xml</td><td/><td/></tr>
<tr id="samjmx"><th><a href="SamJmx.html" title="SamJmx">samjmx</a></th><td>Monitor/interrupt/break a BAM/SAM stream with java JMX http://www.oracle.com/technetwork/articles/java/javamanagement-140525.html</td><td>sam bam jmx monitoring</td><td/><td/></tr>
<tr id="ngsfilessummary"><th><a href="NgsFilesSummary.html" title="NgsFilesSummary">ngsfilessummary</a></th><td>Scan folders and generate a summary of the files (SAMPLE/BAM SAMPLE/VCF etc..)</td><td/><td/><td/></tr>
<tr id="noemptyvcf"><th><a href="NoEmptyVCF.html" title="NoEmptyVCF">noemptyvcf</a></th><td>If VCF is empty or doesn't exists, create a dummy one</td><td/><td/><td><span style="color:orange;" title="deprecated">Was developped at the time where VEP didn't send an output if there was no variant, just a header in the source vcf.</span>. </td></tr>
<tr id="nozerovariationvcf"><th><a href="NoZeroVariationVCF.html" title="NoZeroVariationVCF">nozerovariationvcf</a></th><td>cat a whole VCF, or, if there is no variant, creates a fake one</td><td/><td/><td/></tr>
<tr id="pademptyfastq"><th><a href="PadEmptyFastq.html" title="PadEmptyFastq">pademptyfastq</a></th><td>Pad empty fastq sequence/qual with N/#</td><td>fastq</td><td/><td><span style="color:orange;" title="deprecated">use awk</span>. </td></tr>
<tr id="pubmeddump"><th><a href="PubmedDump.html" title="PubmedDump">pubmeddump</a></th><td>Dump XML results from pubmed/Eutils</td><td>ncbi pubmed xml</td><td/><td/></tr>
<tr id="pubmedorcidgraph"><th><a href="PubmedOrcidGraph.html" title="PubmedOrcidGraph">pubmedorcidgraph</a></th><td>Creates a graph from Pubmed and Authors' Orcid identifiers</td><td>pubmed ncbi orcid</td><td/><td/></tr>
<tr id="pubmedfilterjs"><th><a href="PubmedFilterJS.html" title="PubmedFilterJS">pubmedfilterjs</a></th><td>Filters Pubmed XML with a javascript  (java rhino) expression. Context contain 'article' a  PubmedBookArticle or a PubmedArticle and 'index', the index in the XML file.</td><td>pubmed javascript xml ncbi</td><td/><td/></tr>
<tr id="referencetovcf"><th><a href="ReferenceToVCF.html" title="ReferenceToVCF">referencetovcf</a></th><td>Creates a VCF containing all the possible substitutions from a Reference Genome.</td><td>vcf reference fasta</td><td/><td/></tr>
<tr id="sam2json"><th><a href="SamToJson.html" title="SamToJson">sam2json</a></th><td>Convert a SAM input to JSON</td><td>sam bam json</td><td/><td/></tr>
<tr id="sam2psl"><th><a href="SamToPsl.html" title="SamToPsl">sam2psl</a></th><td>Convert SAM/BAM to PSL http://genome.ucsc.edu/FAQ/FAQformat.html#format2 or BED12</td><td/><td/><td/></tr>
<tr id="sam2tsv"><th><a href="Sam2Tsv.html" title="Sam2Tsv">sam2tsv</a></th><td>Prints the SAM alignments as a TAB delimited file.</td><td>sam bam table tsv</td><td/><td/></tr>
<tr id="sam4weblogo"><th><a href="SAM4WebLogo.html" title="SAM4WebLogo">sam4weblogo</a></th><td>Sequence logo for different alleles or generated from SAM/BAM </td><td>sam bam visualization logo</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="Samclipindelfraction"><th><a href="SamClipIndelFraction.html" title="SamClipIndelFraction">Samclipindelfraction</a></th><td>Extract clipping/indel fraction from BAM</td><td/><td/><td/></tr>
<tr id="samextractclip"><th><a href="SamExtractClip.html" title="SamExtractClip">samextractclip</a></th><td>Extract Clipped Sequences from a SAM. Ouput is a FASTQ</td><td/><td/><td/></tr>
<tr id="samfixcigar"><th><a href="SamFixCigar.html" title="SamFixCigar">samfixcigar</a></th><td>Fix Cigar String in SAM replacing 'M' by 'X' or '='</td><td>sam bam cigar</td><td/><td/></tr>
<tr id="samgrep"><th><a href="SamGrep.html" title="SamGrep">samgrep</a></th><td>grep read-names in a bam file</td><td>sam bam</td><td/><td/></tr>
<tr id="samshortinvert"><th><a href="SamShortInvertion.html" title="SamShortInvertion">samshortinvert</a></th><td>Scan short inversions in SAM</td><td/><td>[ID:0000023]it_is_broken</td><td/></tr>
<tr id="samstats01"><th><a href="BamStats01.html" title="BamStats01">samstats01</a></th><td>Statistics about the reads in a BAM.</td><td/><td/><td/></tr>
<tr id="sortvcfoninfo"><th><a href="SortVcfOnInfo.html" title="SortVcfOnInfo">sortvcfoninfo</a></th><td>Sort a VCF a field in the INFO column</td><td>vcf sort annotation</td><td/><td/></tr>
<tr id="sortvcfonref2"><th><a href="SortVcfOnRef2.html" title="SortVcfOnRef2">sortvcfonref2</a></th><td>Sort a VCF using the internal dictionary or an external reference order (Deprecated: use picard SortVcf).</td><td>vcf sort</td><td/><td><span style="color:orange;" title="deprecated">use picard sortvcf</span>. </td></tr>
<tr id="splitbytile"><th><a href="SplitByTile.html" title="SplitByTile">splitbytile</a></th><td>Split Bam By tile</td><td>sam bam</td><td/><td/></tr>
<tr id="vcf2hilbert"><th><a href="VcfToHilbert.html" title="VcfToHilbert">vcf2hilbert</a></th><td>Plot a Hilbert Curve from a VCF file.</td><td>vcf image visualization</td><td/><td/></tr>
<tr id="vcf2postscript"><th><a href="VcfToPostscript.html" title="VcfToPostscript">vcf2postscript</a></th><td>Print VCF context as Postscript</td><td>vcf postscript</td><td/><td/></tr>
<tr id="vcf2rdf"><th><a href="VcfToRdf.html" title="VcfToRdf">vcf2rdf</a></th><td>convert VCF to RDF (N3 notation)</td><td>vcf rdf</td><td/><td/></tr>
<tr id="vcf2sql"><th><a href="VcfToSql.html" title="VcfToSql">vcf2sql</a></th><td>Generate the SQL code to insert a VCF into mysql</td><td>vcf sql</td><td/><td/></tr>
<tr id="vcf2xml"><th><a href="Vcf2Xml.html" title="Vcf2Xml">vcf2xml</a></th><td>Convert VCF to XML</td><td>vcf xml</td><td/><td/></tr>
<tr id="vcfannobam"><th><a href="VCFAnnoBam.html" title="VCFAnnoBam">vcfannobam</a></th><td>Annotate a VCF with the Coverage statistics of a BAM file+  BED file of capture. It uses the Cigar string instead of the start/end to get the voverage</td><td/><td/><td><span style="color:orange;" title="deprecated">useless: use DP/DP4 in the Genotypes</span>. </td></tr>
<tr id="vcfbiomart"><th><a href="VcfBiomart.html" title="VcfBiomart">vcfbiomart</a></th><td>BiomartQueries with VCF</td><td>vcf ensembl biomart</td><td/><td/></tr>
<tr id="vcfcadd"><th><a href="VcfCadd.html" title="VcfCadd">vcfcadd</a></th><td>Annotate VCF with  Combined Annotation Dependent Depletion (CADD) (Kircher &amp; al. A general framework for estimating the relative pathogenicity of human genetic variants. Nat Genet. 2014 Feb 2. doi: 10.1038/ng.2892.PubMed PMID: 24487276.</td><td/><td/><td/></tr>
<tr id="vcfcmppred"><th><a href="VCFComparePredictions.html" title="VCFComparePredictions">vcfcmppred</a></th><td>Compare predictions (SNPEff, VEP) for several VCFs</td><td/><td/><td/></tr>
<tr id="vcfcomm"><th><a href="VCFComm.html" title="VCFComm">vcfcomm</a></th><td>Equivalent of linux comm for VCF</td><td/><td/><td/></tr>
<tr id="vcfcompare"><th><a href="VCFCompare.html" title="VCFCompare">vcfcompare</a></th><td>Compares two VCF files</td><td>vcf compare</td><td/><td/></tr>
<tr id="vcfcomparegt"><th><a href="VCFCompareGT.html" title="VCFCompareGT">vcfcomparegt</a></th><td> compare two or more genotype-callers for the same individuals. Produce a VCF with FORMAT fields indicating if a genotype is new or modified.</td><td>vcf compare</td><td/><td/></tr>
<tr id="vcfconcat"><th><a href="VcfConcat.html" title="VcfConcat">vcfconcat</a></th><td>Concatenante sorted VCF with same sample, does NOT merge genotypes</td><td>vcf</td><td/><td/></tr>
<tr id="vcfcutsamples"><th><a href="VcfCutSamples.html" title="VcfCutSamples">vcfcutsamples</a></th><td>Select/Exclude some samples from a VCF</td><td>vcf sample</td><td/><td><span style="color:orange;" title="deprecated">use bcftools or gatk SelectVariants</span>. </td></tr>
<tr id="vcffilterdoid"><th><a href="VcfFilterDoid.html" title="VcfFilterDoid">vcffilterdoid</a></th><td>Set Filters for VCF annotated with SNPEFF or VEP tesing if Genes are part / are not part of a GeneOntology tree. </td><td/><td/><td/></tr>
<tr id="vcfjaspar"><th><a href="VcfJaspar.html" title="VcfJaspar">vcfjaspar</a></th><td>Finds JASPAR profiles in VCF</td><td>vcf matrix jaspar</td><td/><td/></tr>
<tr id="vcfliftover"><th><a href="VcfLiftOver.html" title="VcfLiftOver">vcfliftover</a></th><td>Lift-over a VCF file</td><td>vcf liftover</td><td/><td/></tr>
<tr id="vcfmerge"><th><a href="VCFMerge2.html" title="VCFMerge2">vcfmerge</a></th><td>Merge VCF Files</td><td>vcf sort</td><td/><td><span style="color:orange;" title="deprecated">use GATK combineVariants.</span>. </td></tr>
<tr id="vcfmulti2one"><th><a href="VcfMultiToOne.html" title="VcfMultiToOne">vcfmulti2one</a></th><td>&gt;Convert VCF with multiple samples to a VCF with one SAMPLE, duplicating variant and adding the sample name in the INFO column</td><td>vcf sample</td><td/><td/></tr>
<tr id="vcfregulomedb"><th><a href="VcfRegulomeDB.html" title="VcfRegulomeDB">vcfregulomedb</a></th><td>Annotate a VCF with the Regulome data (http://regulome.stanford.edu/</td><td/><td/><td/></tr>
<tr id="vcfrenamechr"><th><a href="ConvertVcfChromosomes.html" title="ConvertVcfChromosomes">vcfrenamechr</a></th><td>Convert the names of the chromosomes in a VCF file</td><td>vcf contig chromosome convert</td><td/><td/></tr>
<tr id="vcfrenamesamples"><th><a href="VcfRenameSamples.html" title="VcfRenameSamples">vcfrenamesamples</a></th><td>Rename the Samples in a VCF</td><td>vcf sample</td><td/><td/></tr>
<tr id="vcfresetvcf"><th><a href="VcfRemoveGenotypeIfInVcf.html" title="VcfRemoveGenotypeIfInVcf">vcfresetvcf</a></th><td>Reset Genotypes in VCF (./.) if they've been found in another VCF indexed with tabix</td><td>vcf genotype</td><td/><td/></tr>
<tr id="vcfsetdict"><th><a href="VcfSetSequenceDictionary.html" title="VcfSetSequenceDictionary">vcfsetdict</a></th><td>Set the `##contig` lines in a VCF header on the fly</td><td>vcf dict fai</td><td/><td/></tr>
<tr id="vcfshuffle"><th><a href="VCFShuffle.html" title="VCFShuffle">vcfshuffle</a></th><td>Shuffle a VCF</td><td>vcf</td><td/><td/></tr>
<tr id="vcfSimulator"><th><a href="VcfSimulator.html" title="VcfSimulator">vcfSimulator</a></th><td>Generate a VCF</td><td/><td/><td/></tr>
<tr id="vcfstats"><th><a href="VcfStats.html" title="VcfStats">vcfstats</a></th><td>Produce VCF statitics</td><td>vcf stats burden gnuplot</td><td/><td/></tr>
<tr id="vcfcombinetwosnvs"><th><a href="VCFCombineTwoSnvs.html" title="VCFCombineTwoSnvs">vcfcombinetwosnvs</a></th><td>Detect Mutations than are the consequences of two distinct variants. This kind of variant might be ignored/skipped from classical variant consequence predictor. Idea from @SolenaLS and then @AntoineRimbert</td><td>vcf annotation prediction protein</td><td/><td/></tr>
<tr id="vcfstripannot"><th><a href="VCFStripAnnotations.html" title="VCFStripAnnotations">vcfstripannot</a></th><td>Removes one or more field from the INFO/FORMAT column of a VCF.</td><td>vcf</td><td/><td><span style="color:orange;" title="deprecated">Use bcftools annotate -x </span>. </td></tr>
<tr id="vcftabixml"><th><a href="VCFTabixml.html" title="VCFTabixml">vcftabixml</a></th><td> annotate a value from a vcf+xml file</td><td>vcf xml</td><td/><td/></tr>
<tr id="vcfvcf"><th><a href="VcfVcf.html" title="VcfVcf">vcfvcf</a></th><td>Get the INFO from a VCF and use it for another VCF</td><td/><td/><td><span style="color:orange;" title="deprecated">obsolete. use GATK</span>. </td></tr>
<tr id="worldmapgenome"><th><a href="WorldMapGenome.html" title="WorldMapGenome">worldmapgenome</a></th><td>Genome/Map of the world. Input is a BED file: chrom/start/end/country.</td><td>gis</td><td>[ID:0000017]just_a_proof_of_concept</td><td/></tr>
<tr id="uniprotfilterjs"><th><a href="UniprotFilterJS.html" title="UniprotFilterJS">uniprotfilterjs</a></th><td>Filters Uniprot DUMP+ XML with a javascript  (java rhino) expression. Context contain 'entry' an uniprot entry and 'index', the index in the XML file.</td><td>unitprot javascript xjc xml</td><td/><td/></tr>
<tr id="skipxmlelements"><th><a href="SkipXmlElements.html" title="SkipXmlElements">skipxmlelements</a></th><td>Filter XML elements with a javascript  (java rhino) expression. Context contain 'element' the current element. It implementsthe interface Tag described in  SkipXmlElements.class</td><td>xml javascript</td><td/><td/></tr>
<tr id="vcfensemblvep"><th><a href="VcfEnsemblVepRest.html" title="VcfEnsemblVepRest">vcfensemblvep</a></th><td>Annotate a VCF with ensembl REST API</td><td>vcf annotation rest ensembl</td><td/><td/></tr>
<tr id="vcfgroupbypop"><th><a href="VcfGroupByPopulation.html" title="VcfGroupByPopulation">vcfgroupbypop</a></th><td>Group VCF data by population, creates a VCF  where each 'SAMPLE' is a population</td><td/><td/><td/></tr>
<tr id="bamtile"><th><a href="BamTile.html" title="BamTile">bamtile</a></th><td>Answer to @sjackman : Is there a bedtools command to determine a minimal tiling path? A minimal set of features that cover a maximum of the target.</td><td>bam sam</td><td/><td/></tr>
<tr id="xcontaminations"><th><a href="XContaminations.html" title="XContaminations">xcontaminations</a></th><td>For @AdrienLeger2 : cross contamination between samples in same lane</td><td/><td/><td/></tr>
<tr id="biostar3654"><th><a href="Biostar3654.html" title="Biostar3654">biostar3654</a></th><td>show blast alignment with annotations</td><td>blast xml annotation</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfburden"><th><a href="VcfBurden.html" title="VcfBurden">vcfburden</a></th><td>Solena: vcf to (chrom/pos/ref/alt/individus(G:0/1/2/-9)</td><td/><td/><td><span style="color:orange;" title="deprecated">deprecated</span>. </td></tr>
<tr id="vcfreplacetag"><th><a href="VCFReplaceTag.html" title="VCFReplaceTag">vcfreplacetag</a></th><td>Replace the key for INFO/FORMAT/FILTER</td><td>vcf</td><td/><td/></tr>
<tr id="vcfpeekvcf"><th><a href="VcfPeekVcf.html" title="VcfPeekVcf">vcfpeekvcf</a></th><td>Get the INFO from a VCF and use it for another VCF</td><td>vcf annotation</td><td/><td/></tr>
<tr id="vcfgetvariantbyIndex"><th><a href="VcfGetVariantByIndex.html" title="VcfGetVariantByIndex">vcfgetvariantbyIndex</a></th><td>Access a Plain or BGZF-compressed VCF file by index</td><td>vcf</td><td/><td/></tr>
<tr id="vcfmulti2oneinfo"><th><a href="VcfMultiToOneInfo.html" title="VcfMultiToOneInfo">vcfmulti2oneinfo</a></th><td>'one INFO with N values' to 'N variants with one INFO'</td><td>vcf</td><td/><td/></tr>
<tr id="bedindextabix"><th><a href="BedIndexTabix.html" title="BedIndexTabix">bedindextabix</a></th><td>Index and sort a Bed on the fly with Tabix.</td><td>bed tabix</td><td/><td/></tr>
<tr id="vcf2bam"><th><a href="VcfToBam.html" title="VcfToBam">vcf2bam</a></th><td>vcf to bam</td><td>ref vcf bam</td><td/><td/></tr>
<tr id="vcffilterxpath"><th><a href="VcfFilterXPath.html" title="VcfFilterXPath">vcffilterxpath</a></th><td>Filter a VCF with a XPATH expression on a INFO tag containing a base64 encodede xml document</td><td>vcf xml xpath</td><td>[ID:0000017]just_a_proof_of_concept</td><td/></tr>
<tr id="biostar140111"><th><a href="Biostar140111.html" title="Biostar140111">biostar140111</a></th><td>How to obtain human genotype data from dpSNP ftp?</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="extendrefwithreads"><th><a href="ExtendReferenceWithReads.html" title="ExtendReferenceWithReads">extendrefwithreads</a></th><td>Extending ends of sequences with the help of reads</td><td>read fastq reference sam bam</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="pcrslicereads"><th><a href="PcrSliceReads.html" title="PcrSliceReads">pcrslicereads</a></th><td>Mark PCR reads to their PCR amplicon</td><td>pcr sam bam cigar</td><td/><td/></tr>
<tr id="vcfjmx"><th><a href="VcfJmx.html" title="VcfJmx">vcfjmx</a></th><td>Monitor/interrupt/break a VCF stream with java JMX http://www.oracle.com/technetwork/articles/java/javamanagement-140525.html</td><td>java jmx vcf</td><td/><td/></tr>
<tr id="gtf2xml"><th><a href="Gtf2Xml.html" title="Gtf2Xml">gtf2xml</a></th><td>Convert GTF to XML</td><td>xml gtf</td><td/><td/></tr>
<tr id="sortsamrefname"><th><a href="SortSamRefName.html" title="SortSamRefName">sortsamrefname</a></th><td>Sort a BAM of contig and then on name</td><td>sam sort</td><td/><td/></tr>
<tr id="biostar154220"><th><a href="Biostar154220.html" title="Biostar154220">biostar154220</a></th><td>Cap BAM to a given coverage</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar160470"><th><a href="Biostar160470.html" title="Biostar160470">biostar160470</a></th><td>Getting untranslated nucleotide sequences on tblastn standalone </td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar165777"><th><a href="Biostar165777.html" title="Biostar165777">biostar165777</a></th><td>Split a XML file</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfcomparecallers"><th><a href="VcfCompareCallers.html" title="VcfCompareCallers">vcfcomparecallers</a></th><td>Compare two VCFs and print common/exclusive information for each sample/genotype</td><td>vcf compare genotype</td><td/><td/></tr>
<tr id="bamclip2insertion"><th><a href="BamClipToInsertion.html" title="BamClipToInsertion">bamclip2insertion</a></th><td>Convert SOFT clip to Insertion of other read confirm it</td><td>sam bam clip</td><td/><td/></tr>
<tr id="localrealignreads"><th><a href="LocalRealignReads.html" title="LocalRealignReads">localrealignreads</a></th><td>Local Realignment of Reads</td><td/><td/><td/></tr>
<tr id="biostar170742"><th><a href="Biostar170742.html" title="Biostar170742">biostar170742</a></th><td>convert sam format to axt Format</td><td>sam axt</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar172515"><th><a href="Biostar172515.html" title="Biostar172515">biostar172515</a></th><td>Convert BAI to XML</td><td>bai bam xml</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar173114"><th><a href="Biostar173114.html" title="Biostar173114">biostar173114</a></th><td>make a bam file smaller by removing unwanted information see also https://www.biostars.org/p/173114/</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar175929"><th><a href="Biostar175929.html" title="Biostar175929">biostar175929</a></th><td>Construct a combination set of fasta sequences from a vcf see also https://www.biostars.org/p/175929/</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfcalledwithanothermethod"><th><a href="VcfCalledWithAnotherMethod.html" title="VcfCalledWithAnotherMethod">vcfcalledwithanothermethod</a></th><td>Compare one vcf with other , add a flag to tell if a variant was called with another method. Vcf must be sorted on the same Dict.</td><td>vcf compare concordance</td><td/><td/></tr>
<tr id="biostar178713"><th><a href="Biostar178713.html" title="Biostar178713">biostar178713</a></th><td>split bed file into several bed files where each region is separated of any other by N bases https://www.biostars.org/p/178713/</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfremovegenotypejs"><th><a href="VcfRemoveGenotypeJs.html" title="VcfRemoveGenotypeJs">vcfremovegenotypejs</a></th><td>Reset Genotype in VCF using a javascript expression</td><td/><td/><td/></tr>
<tr id="bamstats02"><th><a href="BamStats02.html" title="BamStats02">bamstats02</a></th><td>Statistics about the flags and reads in a BAM</td><td/><td/><td/></tr>
<tr id="bamstats02view"><th><a href="BamStats02View.html" title="BamStats02View">bamstats02view</a></th><td>Statistics about the flags and reads in a BAM. Visualize date from BamStats02</td><td/><td/><td/></tr>
<tr id="biostar105754"><th><a href="Biostar105754.html" title="Biostar105754">biostar105754</a></th><td>bigwig : peak distance from specific genomic region</td><td>wig bigwig</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="gff2knowngene"><th><a href="Gff2KnownGene.html" title="Gff2KnownGene">gff2knowngene</a></th><td>Convert GFF3 format to UCSC knownGene format.</td><td>gff knownGene ucsc convert</td><td/><td/></tr>
<tr id="bam2sql"><th><a href="BamToSql.html" title="BamToSql">bam2sql</a></th><td>Convert a SAM/BAM to sqlite statements</td><td>bam sam sql sqlite</td><td/><td/></tr>
<tr id="vcf2zip"><th><a href="VcfToZip.html" title="VcfToZip">vcf2zip</a></th><td>Reads a stream of concatenated VCFs and insert them into a Zip file</td><td/><td/><td/></tr>
<tr id="pubmedgender"><th><a href="PubmedGender.html" title="PubmedGender">pubmedgender</a></th><td>Add gender-related attributes in the Author tag of pubmed xml. </td><td>pubmed gender ncbi xml</td><td/><td/></tr>
<tr id="pubmedmap"><th><a href="PubmedMap.html" title="PubmedMap">pubmedmap</a></th><td>Use Pubmed Author's Affiliation to map the authors .</td><td>pubmed xml gis map</td><td/><td/></tr>
<tr id="splitvcf"><th><a href="SplitVcf.html" title="SplitVcf">splitvcf</a></th><td>split a vcf...</td><td>vcf</td><td/><td/></tr>
<tr id="forkvcf"><th><a href="GBrowserHtml.html" title="GBrowserHtml">forkvcf</a></th><td>Fork a VCF.</td><td/><td/><td/></tr>
<tr id="bim2vcf"><th><a href="BimToVcf.html" title="BimToVcf">bim2vcf</a></th><td>convert a .bim to a .vcf</td><td/><td/><td/></tr>
<tr id="concatsam"><th><a href="ConcatSam.html" title="ConcatSam">concatsam</a></th><td>concat sam files</td><td>sam bam</td><td/><td/></tr>
<tr id="samreadlengthdistribution"><th><a href="SamReadLengthDistribution.html" title="SamReadLengthDistribution">samreadlengthdistribution</a></th><td>Sam read length distribution</td><td>sam bam histogram</td><td/><td/></tr>
<tr id="biostar214299"><th><a href="Biostar214299.html" title="Biostar214299">biostar214299</a></th><td>Extract allele specific reads from bamfiles</td><td>sam bam variant snp</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="cmpbams4"><th><a href="CompareBams4.html" title="CompareBams4">cmpbams4</a></th><td>Compare two BAM files. Print a tab-delimited report</td><td>sam bam compare</td><td/><td/></tr>
<tr id="biostar234081"><th><a href="Biostar234081.html" title="Biostar234081">biostar234081</a></th><td>convert extended CIGAR to regular CIGAR</td><td>sam bam cigar</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfgnomad"><th><a href="VcfGnomad.html" title="VcfGnomad">vcfgnomad</a></th><td>Peek annotations from gnomad</td><td>vcf annotation gnomad</td><td/><td/></tr>
<tr id="vcf2svg"><th><a href="VcfToSvg.html" title="VcfToSvg">vcf2svg</a></th><td>write a vcf to svg , with gene context</td><td>vcf svg xlm visualization</td><td/><td/></tr>
<tr id="mergeblastxml"><th><a href="MergeBlastXml.html" title="MergeBlastXml">mergeblastxml</a></th><td>merge XML blast results (same Iteration/Iteration_query-def in multiple xml files</td><td>blast xml</td><td/><td/></tr>
<tr id="vcfannotwithbeacon"><th><a href="VcfAnnotWithBeacon.html" title="VcfAnnotWithBeacon">vcfannotwithbeacon</a></th><td>Annotate a VCF with ga4gh beacon</td><td>ga4gh beacon vcf annotation</td><td>[ID:0000017]just_a_proof_of_concept</td><td/></tr>
<tr id="commbams"><th><a href="CommBams.html" title="CommBams">commbams</a></th><td>Equivalent of unix 'comm' for bams sorted on queryname</td><td>sam bam comm compare</td><td/><td/></tr>
<tr id="samscansplitreads"><th><a href="SamScanSplitReads.html" title="SamScanSplitReads">samscansplitreads</a></th><td>scan split reads</td><td>sam sv splitreads</td><td/><td/></tr>
<tr id="samretrieveseqandqual"><th><a href="SamRetrieveSeqAndQual.html" title="SamRetrieveSeqAndQual">samretrieveseqandqual</a></th><td>I have a query-sorted BAM file without read/qual sequences and a FASTQ file with the read/qual sequences. Is there a tool to add seq to BAM?  for @sjackman https://twitter.com/sjackman/status/575368165531611136</td><td/><td/><td/></tr>
<tr id="biostar234230"><th><a href="Biostar234230.html" title="Biostar234230">biostar234230</a></th><td>Sliding Window : discriminate partial and fully contained fragments (from a bam file)</td><td/><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="biostar251649"><th><a href="Biostar251649.html" title="Biostar251649">biostar251649</a></th><td> Annotating the flanking bases of SNPs in a VCF file</td><td>vcf annotation sequence reference</td><td>[ID:0000015]i_wrote_it_for_biostars</td><td/></tr>
<tr id="vcfbedsetfilter"><th><a href="VCFBedSetFilter.html" title="VCFBedSetFilter">vcfbedsetfilter</a></th><td>Set FILTER for VCF if it doesn't intersects with BED.</td><td>vcf bed filter</td><td/><td><span style="color:orange;" title="deprecated">use GATK FilterVariants</span>. </td></tr>
<tr id="bamliftover"><th><a href="BamLiftOver.html" title="BamLiftOver">bamliftover</a></th><td>Lift-over a BAM file.</td><td>bam liftover</td><td/><td/></tr>
<tr id="blast2sam"><th><a href="BlastToSam.html" title="BlastToSam">blast2sam</a></th><td>Convert a **BLASTN-XML** input to SAM</td><td>sam blast</td><td/><td/></tr>
<tr id="reduceblast"><th><a href="ReduceBlast.html" title="ReduceBlast">reduceblast</a></th><td>Reduce the size of XML blast, by removing iterations that have no Hit</td><td>blast xml</td><td/><td/></tr>
<tr id="blasn2snp"><th><a href="BlastNToSnp.html" title="BlastNToSnp">blasn2snp</a></th><td>print indel/mismatch in a blastn stream</td><td>blast snp</td><td/><td/></tr>
<tr id="splitbam3"><th><a href="SplitBam3.html" title="SplitBam3">splitbam3</a></th><td>Split a BAM by chromosome group</td><td/><td/><td/></tr>
<tr id="vcfindextabix"><th><a href="VcfIndexTabix.html" title="VcfIndexTabix">vcfindextabix</a></th><td>Index and sort a VCF on the fly with Tabix</td><td>vcf tabix</td><td/><td/></tr>
<tr id="blastfilterjs"><th><a href="BlastFilterJS.html" title="BlastFilterJS">blastfilterjs</a></th><td>Filters a BlastOutput with a javascript expression. The script injects each &lt;Hit&gt; as the variable 'blasthit'. The user script should return 'true' to keep the hit.</td><td>blast js javascript filter</td><td/><td/></tr>
<tr id="SamSlop"><th><a href="SamSlop.html" title="SamSlop">SamSlop</a></th><td>extends sam by 'x' bases</td><td/><td/><td/></tr>
<tr id="vcfgenesplitter"><th><a href="VcfGeneSplitter.html" title="VcfGeneSplitter">vcfgenesplitter</a></th><td>Split VCF+VEP by gene.</td><td/><td/><td/></tr>
<tr id="sammaskalignedbases"><th><a href="SamMaskAlignedBases.html" title="SamMaskAlignedBases">sammaskalignedbases</a></th><td>Mask bases aligned on Reference.</td><td/><td/><td/></tr>
<tr id="queue2make"><th><a href="QueueToMake.html" title="QueueToMake">queue2make</a></th><td>Convert Broad/Queue genomestrip Log stream to Makefile.</td><td/><td/><td/></tr>
<tr id="pcrclipreads"><th><a href="PcrClipReads.html" title="PcrClipReads">pcrclipreads</a></th><td>Soft clip bam files based on PCR target regions</td><td>sam bam pcr bed</td><td/><td/></tr>
<tr id="vcfeigen"><th><a href="VcfEigen01.html" title="VcfEigen01">vcfeigen</a></th><td>Annotator for the data of https://xioniti01.u.hpc.mssm.edu/v1.1/ : Eigen makes use of a variety of functional annotations in both coding and noncoding regions (such as made available by the ENCODE and Roadmap Epigenomics projects), and combines them into one single measure of functional importance.</td><td>vcf annotation variant</td><td/><td/></tr>
<tr id="vcfmovefilterstoinfo"><th><a href="VcfMoveFiltersToInfo.html" title="VcfMoveFiltersToInfo">vcfmovefilterstoinfo</a></th><td>Move any FILTER to the INFO column. reset FILTER to PASS</td><td>vcf burden format info</td><td/><td/></tr>
<tr id="vcfburdenmaf"><th><a href="VcfBurdenMAF.html" title="VcfBurdenMAF">vcfburdenmaf</a></th><td>Burden : MAF for Cases / Controls </td><td>vcf burden maf case control</td><td/><td/></tr>
<tr id="vcfburdenfisherh"><th><a href="VcfBurdenFisherH.html" title="VcfBurdenFisherH">vcfburdenfisherh</a></th><td>Fisher Case /Controls per Variant</td><td/><td/><td/></tr>
<tr id="vcfinjectpedigree"><th><a href="VcfInjectPedigree.html" title="VcfInjectPedigree">vcfinjectpedigree</a></th><td>Injects a pedigree (.ped) file in the VCF header</td><td>vcf pedigree burden</td><td/><td/></tr>
<tr id="samcolortag"><th><a href="SamColorTag.html" title="SamColorTag">samcolortag</a></th><td>Add the UCSC 'YC' color tag in a BAM. See http://software.broadinstitute.org/software/igv/book/export/html/6 and http://genome.ucsc.edu/goldenPath/help/hgBamTrackHelp.html</td><td>sam bam metadata javascript igv visualization</td><td/><td/></tr>
<tr id="vcffilternotinpedigree"><th><a href="VcfFilterNotInPedigree.html" title="VcfFilterNotInPedigree">vcffilternotinpedigree</a></th><td>Adds a FILTER NotInPedigree if the only not(homref) genotypes are not in a pedigree</td><td>burden vcf pedigree</td><td/><td/></tr>
<tr id="vcfburdenexac"><th><a href="VcfBurdenFilterExac.html" title="VcfBurdenFilterExac">vcfburdenexac</a></th><td>Burden filter 3 - Exac</td><td>vcf burden exac</td><td/><td/></tr>
<tr id="vcf2table"><th><a href="VcfToTable.html" title="VcfToTable">vcf2table</a></th><td>convert a vcf to a table, to ease display in the terminal</td><td>vcf table visualization</td><td/><td/></tr>
<tr id="vcfburdenfiltergenes"><th><a href="VcfBurdenFilterGenes.html" title="VcfBurdenFilterGenes">vcfburdenfiltergenes</a></th><td>Filter VEP Output from a list of genes.</td><td>gene vcf vep</td><td/><td/></tr>
<tr id="vcfrefbase"><th><a href="VcfRebase.html" title="VcfRebase">vcfrefbase</a></th><td>Restriction sites overlaping variations in a vcf</td><td>vcf rebase restriction enzyme</td><td/><td/></tr>
<tr id="projectserver"><th><a href="ProjectServer.html" title="ProjectServer">projectserver</a></th><td>Jetty Based http server serving VCF, BAM files</td><td/><td/><td/></tr>
<tr id="vcfwindowvariants"><th><a href="VariantsInWindow.html" title="VariantsInWindow">vcfwindowvariants</a></th><td>Annotate Number of Variants overlaping a sliding window.</td><td>vcf annotation</td><td/><td/></tr>
<tr id="allelefreqcalc"><th><a href="AlleleFrequencyCalculator.html" title="AlleleFrequencyCalculator">allelefreqcalc</a></th><td>Allele Frequency Calculator</td><td>vcf af</td><td/><td><span style="color:orange;" title="deprecated">Use bioalcidae</span>. </td></tr>
<tr id="bamindexreadnames"><th><a href="BamIndexReadNames.html" title="BamIndexReadNames">bamindexreadnames</a></th><td>Build a dictionary of read names to be searched with BamQueryReadNames</td><td/><td/><td/></tr>
<tr id="bamtreepack"><th><a href="BamTreePack.html" title="BamTreePack">bamtreepack</a></th><td>Create a TreeMap from one or more SAM/BAM file. Ouput is a SVG file.</td><td>bam treepack</td><td/><td/></tr>
<tr id="fastqrecordtreepack"><th><a href="FastqRecordTreePack.html" title="FastqRecordTreePack">fastqrecordtreepack</a></th><td>Create a TreeMap from one or more Fastq file. Ouput is a SVG file</td><td/><td/><td/></tr>
<tr id="samfindclippedregions"><th><a href="SamFindClippedRegions.html" title="SamFindClippedRegions">samfindclippedregions</a></th><td/><td/><td/><td/></tr>
<tr id="vcftreepack"><th><a href="VcfTreePack.html" title="VcfTreePack">vcftreepack</a></th><td>Create a TreeMap from one or more VCF. Ouput is a SVG file.</td><td/><td/><td/></tr>
<tr id="vcfburdensplitter"><th><a href="VcfBurdenSplitter.html" title="VcfBurdenSplitter">vcfburdensplitter</a></th><td>Split VCF Using a Burden Splitter (by gene, transcript, etc..)</td><td/><td/><td/></tr>
<tr id="vcfburdenfisherv"><th><a href="VcfBurdenFisherV.html" title="VcfBurdenFisherV">vcfburdenfisherv</a></th><td>Fisher Case / Controls per Variant (Vertical)</td><td>vcf burden fisher</td><td/><td/></tr>
<tr id="vcfburdenrscriptv"><th><a href="VcfBurdenRscriptV.html" title="VcfBurdenRscriptV">vcfburdenrscriptv</a></th><td>Fisher Case / Controls per Variant (Vertical)</td><td>vcf burden fisher</td><td/><td/></tr>
<tr id="vcfderby01"><th><a href="VcfDerby01.html" title="VcfDerby01">vcfderby01</a></th><td>Insert similar VCFs into an Apache Derby Database</td><td>vcf sql derby burden</td><td/><td/></tr>
<tr id="vcfdoest"><th><a href="VcfDoest.html" title="VcfDoest">vcfdoest</a></th><td>generate Transcript information for DOEST test</td><td>vcf burden doest</td><td/><td/></tr>
<tr id="bioalcidae"><th><a href="BioAlcidae.html" title="BioAlcidae">bioalcidae</a></th><td>javascript version of awk for bioinformatics</td><td>sam bam vcf javascript js nashorn</td><td/><td/></tr>
<tr id="picardmetrics2xml"><th><a href="PicardMetricsToXML.html" title="PicardMetricsToXML">picardmetrics2xml</a></th><td>transforms a picard metrics file to XML. See http://plindenbaum.blogspot.fr/2013/02/making-use-of-picard-metrics-files.html</td><td>picard xml metrics</td><td/><td/></tr>
<tr id="casectrlcanvas"><th><a href="CaseControlCanvas.html" title="CaseControlCanvas">casectrlcanvas</a></th><td>draw a chart of case/control maf from a stream of X/Y values</td><td>vcf case control visualization jfx chart maf</td><td/><td/></tr>
<tr id="lowresbam2raster"><th><a href="LowResBam2Raster.html" title="LowResBam2Raster">lowresbam2raster</a></th><td>Low Resolution BAM to raster graphics</td><td>bam alignment graphics visualization png knowngene</td><td/><td/></tr>
<tr id=""><th><a href="GATKCodeGenerator.html" title="GATKCodeGenerator"/></th><td>Gatk code generator</td><td/><td/><td/></tr>
<tr id="knime2txt"><th><a href="KnimeToText.html" title="KnimeToText">knime2txt</a></th><td>converts a Knime Workflow to a html representation.</td><td>knime workflow convert</td><td/><td/></tr>
<tr id="lumpyvcf2circos"><th><a href="LumpyVcfToCircos.html" title="LumpyVcfToCircos">lumpyvcf2circos</a></th><td>Lumpy to Circos</td><td>lumpy circos sv vcf</td><td/><td/></tr>
<tr id="genscan"><th><a href="GenScan.html" title="GenScan">genscan</a></th><td>Paint a Genome Scan picture from a Tab delimited file (CHROM/POS/VALUE1/VALUE2/....).</td><td>chromosome reference chart visualization</td><td/><td/></tr>
<tr id="vcfucsc"><th><a href="VcfUcsc.html" title="VcfUcsc">vcfucsc</a></th><td>annotate an VCF with mysql UCSC data</td><td>ucsc mysql vcf</td><td/><td/></tr>
<tr id="xsltstream"><th><a href="XsltStream.html" title="XsltStream">xsltstream</a></th><td>XSLT transformation for large XML files. xslt is only applied on a given subset of nodes.</td><td>xml xslt xsl stylesheet</td><td/><td/></tr>
<tr id="splitread"><th><a href="SplitRead.html" title="SplitRead">splitread</a></th><td>TODO</td><td>sam bam</td><td/><td/></tr>
<tr id="vcfloopovergenes"><th><a href="VcfLoopOverGenes.html" title="VcfLoopOverGenes">vcfloopovergenes</a></th><td>Generates a BED file of the Genes in an annotated VCF, loop over those genes and generate a VCF for each gene, then execute an optional command.</td><td>vcf gene burden</td><td/><td/></tr>
<tr id="vcffilterjdk"><th><a href="VcfFilterJdk.html" title="VcfFilterJdk">vcffilterjdk</a></th><td>Filtering VCF with in-memory-compiled java expressions</td><td>vcf filter java jdk</td><td/><td/></tr>
<tr id="optimizer"><th><a href="Optimizer.html" title="Optimizer">optimizer</a></th><td>Genetic-Programming-like parameters optimizer</td><td>genetic-programming</td><td/><td/></tr>
<tr id="bioalcidaejdk"><th><a href="BioAlcidaeJdk.html" title="BioAlcidaeJdk">bioalcidaejdk</a></th><td>java-based version of awk for bioinformatics</td><td>sam bam vcf javascript jdk</td><td/><td/></tr>
<tr id="copynumber01"><th><a href="CopyNumber01.html" title="CopyNumber01">copynumber01</a></th><td>experimental CNV detection. Doesn't work for now.</td><td>cnv bam sam</td><td/><td/></tr>
<tr id="samjdk"><th><a href="SamJdk.html" title="SamJdk">samjdk</a></th><td>Filters a BAM using a java expression compiled in memory.</td><td>sam bam java jdk filter</td><td/><td/></tr>
</tbody>

</table>


<h2>Other Pages</h2>
<table>
<thead>
<tr>
		
		<th>Name</th>
		<th>Description</th>
		<th>Keywords</th>
		<th>Terms</th>
		<th>Misc</th>
</tr>
</thead>
<tbody>
	<tr><th><a href="KnimeIntegration.html" title="LowResBam2Raster">Knime Integration</a></th><td>How to use HTSJDK/Jvarkit in knime</td><td>bam sam knime vcf annotation</td><td/><td/></tr>
</tbody>
</table>
</div>

</body>

</html>